repl/src/interactive_helper.rs

```diff
@@ -33,26 +51,266 @@ impl Highlighter for ReplHelper {
     }
 
     fn highlight_hint<'h>(&self, hint: &'h str) -> Cow<'h, str> {
+        //Render the hints in a lighter font
         Owned("\x1b[2m".to_owned() + hint + "\x1b[m")
     }
 
-    fn highlight<'l>(&self, line: &'l str, pos: usize) -> Cow<'l, str> {
-        self.highlighter.highlight(line, pos)
+    fn highlight<'l>(&self, line: &'l str, _pos: usize) -> Cow<'l, str> {
+
+        //See if we need to highlight the bracket matching the cursor position
+        //BUG: this could possibly get tripped up by parenthesis inside comments and string literals
+        let mut blink_char = None;
+        if let Some((bracket, pos)) = self.cursor_bracket.get() {
+            blink_char = find_matching_bracket(line, pos, bracket);
+        }
+
+        //Iterate over the syntax nodes generated by the parser, coloring them appropriately
+        let mut colored_line = String::with_capacity(line.len() * 2);
+        let mut bracket_depth = 0;
+        self.metta.borrow_mut().inside_env(|metta| {
+            let mut parser = SExprParser::new(line);
```

Comment:
I don't completely follow.  Rustyline is fundamentally a one-line-at-a-time interface. (although lines from the repl's perspective can contain linebreak chars, and the arrow keys behave like you'd hope.

All the code in this `highlight` function works fine with big complex multi-line statement (I pasted in many of the files in `python/tests/scripts/` to test this) although the terminal starts to lag pretty bad when redrawing a massive amount of text on every keystroke.

I think it would be pretty straightforward as a next step to adapt this to a multi-line editor, but I don't know if RustyLine is the best choice.  The UI design of a multi-line editor and a repl are pretty different.

Do you have a particular piece of software in mind, that we could pattern the UI after?

Reply:
I mean when I am trying to input `(a b c<Enter>` I see `> (a b c - Unexpected end of expression
` error instead of cursor on the next line.

Reply:
Python for example allows multiline definitions for functions or classes, thus it is not multiline editor, on the other hand it allows continue code on the next line. Most command line shells allows the same. If it is complex to do with rustyline then I think it is fine for now. Totally fine for this PR.

Reply:
Ah.. So you just mean the ability to type "enter" without it "submitting".  Because you can paste a linefeed char and it does what you describe.

Reply:
Yes

Reply:
Ideally tool should decide whether item is ready to be submitted or should be continued on next "line" before submitting.

Reply:
I want to keep the syntax error behavior in there, because it would be frustrating to the user if it just assumed a line with an error was unfinished, if the user didn't spot the error.

I will try the following behavior:
* if user hits enter and line is valid, it will be submitted.
* if user hits enter and line is invalid, it will treat it as a newline
* If user hits enter twice in a row, it will report a syntax error

Reply:
Usually prompt is changed to give user a hint that it is not completed line. For instance Python prints `...` instead of `>` at the beginning of such line.

Reply:
But I like your idea as well.

Reply:
Ok.  Let me know what you think of the behavior in 452e46f

Reply:
Small tweak in fb13d4f.  If you changed the last char, but the last char wasn't a newline, and the line was still invalid, it would report a syntax error rather than allow the edit.

Reply:
Looks good to me thanks!